Provide descriptive error on invalid table reference

codyjlin opened the pull request:
Signed-off-by: Cody Lin codyl@twitter.com

**What this PR does / why we need it**:
This PR creates a more useful error message when a user specifies an invalid table. It catches it before any other operations, like trying to infer the DataSource `event_timestamp_column`.

**Which issue(s) this PR fixes**:
Fixes #1596

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

woop commented on sdk/python/feast/data_source.py:
```diff
@@ -767,6 +774,17 @@ def get_table_column_names_and_types(self) -> Iterable[Tuple[str, str]]:
+        from google.api_core.exceptions import NotFound
+        from google.cloud import bigquery
+
+        client = bigquery.Client()
```
Should this client receive configuration?

codyjlin replied:
Do you mean like the following?
```python
gcp_project, _ = bq_table_ref.split(":")
bq_client = bigquery.Client(project=gcp_project)
```

I saw in line https://github.com/feast-dev/feast/blob/master/sdk/python/feast/data_source.py#L745 and a lot of other instances where it simply just has `client = bigquery.Client()` without specifying the project, do you know when and when it's not necessary?

mavysavydav replied:
i think calling Client w/o specifying project should be fine. https://googleapis.dev/python/bigquery/latest/_modules/google/cloud/bigquery/client.html#Client.get_table says get_tables uses from_string https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.table.TableReference.html#google.cloud.bigquery.table.TableReference.from_string which parses the string for the project and dataset if a default project id isn’t specified

woop commented on sdk/python/feast/data_source.py:
```diff
@@ -671,6 +671,13 @@ def __init__(
         self._bigquery_options = BigQueryOptions(table_ref=table_ref, query=query)
 
+        if not self._table_exists():
+            raise NameError(
```
Wouldn't it be cleaner to raise an exception in `_table_exists()`?

codyjlin replied:
Good idea, let me make it cleaner.

mavysavydav commented on sdk/python/tests/test_data_source.py:
```diff
@@ -0,0 +1,17 @@
+@pytest.mark.integration
+def test_nonexistent_bq_source():
+    client = bigquery.Client()
+    table_ref = "project.dataset.nonexistent_table"
```
we don't need this variable, can directly place string as param below

achals commented on sdk/python/feast/data_source.py:
```diff
@@ -767,6 +768,21 @@ def get_table_column_names_and_types(self) -> Iterable[Tuple[str, str]]:
 
         return name_type_pairs
 
+    def _table_exists(self):
```
nit, can you rename this `_assert_table_exists` or something?

achals reviewed (commented):
looks good aside from a single nit

achals commented:
/kind bug

achals commented:
/ok-to-test

achals commented:
@codyjlin FYI, you should be able to run `make format`, `make lint` & `make test` locally as well

codyjlin commented:
> @codyjlin FYI, you should be able to run make format, make lint & make test locally as well

@achals Apologies!! I had some trouble with setup where other files that I didn't modify had errored out - I thought it might have just been issues with my configuration vs in CI, so I thought I would just try to have it run on here first. Now I've fixed my setup and can see the same errors locally 👍

codyjlin commented:
@achals Noted in this separate issue! https://github.com/feast-dev/feast/issues/1631#issuecomment-857960751

Looks like I have some other integration test errors - many complaining about invalid project id `rh_prod`, and some relating to my code changes in `data_source.py`. Could I ask for your help on the former case and does it relate to any other configurations I'm missing with gcp/BQ?

achals commented:
/ok-to-test

achals commented:
> @achals Noted in this separate issue! [#1631 (comment)](https://github.com/feast-dev/feast/issues/1631#issuecomment-857960751)
> 
> Looks like I have some other integration test errors - many complaining about invalid project id `rh_prod`, and some relating to my code changes in `data_source.py`. Could I ask for your help on the former case and does it relate to any other configurations I'm missing with gcp/BQ?

Awesome, seems like this PR uncovered some issues in our existing test repos. I'll make some fixes and will ask you to rebase once I'm done

codyjlin commented:
> Awesome, seems like this PR uncovered some issues in our existing test repos. I'll make some fixes and will ask you to rebase once I'm done

@achals Thanks!! I just rebased in case I was missing any other important commits - sorry this PR has become so ugly for such a simple task. I also just fixed the tests relevant to my code changes, so I think that's all I can do for now. Let me know when the other tests might be fixed, and I'll be happy to rebase and finally get my first PR merged. 😃 

For reference, here's the [current error]( https://github.com/feast-dev/feast/pull/1627/checks?check_run_id=2788157237#step:10:235) (x5) I'm getting in the integration tests:
```SHELL
google.api_core.exceptions.BadRequest: 400 GET https://bigquery.googleapis.com/bigquery/v2/projects/rh_prod/datasets/ride_hailing_co/tables/drivers?prettyPrint=false: Invalid project ID 'rh_prod'. Project IDs must contain 6-63 lowercase letters, digits, or dashes. Some project IDs also include domain name separated by a colon. IDs must start with a letter and may not end with a dash.
```

It happens for the following tests:
```SHELL
FAILED sdk/python/tests/test_cli_gcp.py::test_basic - assert 1 == 0
FAILED sdk/python/tests/test_cli_local.py::test_workflow - AssertionError: Ex...
FAILED sdk/python/tests/test_cli_local.py::test_non_local_feature_repo - Asse...
FAILED sdk/python/tests/test_cli_redis.py::test_basic - assert 1 == 0
FAILED sdk/python/tests/test_online_retrieval.py::test_online - AssertionError
FAILED sdk/python/tests/test_partial_apply.py::test_partial - AssertionError
```

achals commented:
FYI @tsotnet is working on updating the tests on our end.

@tsotnet , we need to update any feature repo in our tests that references the `rh_prod` gcp project to something valid, preferrably from https://cloud.google.com/bigquery/public-data

woop commented on sdk/python/feast/data_source.py:
```diff
@@ -767,6 +769,21 @@ def get_table_column_names_and_types(self) -> Iterable[Tuple[str, str]]:
+        except NotFound:
+            raise NameError(
+                f"""
+                Unable to find table {self.table_ref} in BigQuery. Please check that table exists.
```
Just swooping by here. Don't have all the context. But if the idea is that we fail based on an upstream table being missing, then would it make more sense to
* Use a custom error type
* Make sure that our tests test all data sources (BigQuery and File sources) and ensure that they raise this custom exception on a missing source?

codyjlin replied:
Good idea! Let me know if the changes in https://github.com/feast-dev/feast/pull/1627/commits/f15f3f61865e7dadbbdca41a9e6ad755008ef93f are what you meant.

codyjlin replied:
I tried to check FileSources in https://github.com/feast-dev/feast/pull/1627/commits/f15f3f61865e7dadbbdca41a9e6ad755008ef93f but it ended up being more complicated than I had expected - seems like a lot of tests were failing maybe because they use temporary files. Either way, it seems like a larger change that could be externalized in a separate issue/PR.

I did adopt the change for a custom exception though, thanks for the feedback!

woop commented on sdk/python/feast/errors.py:
```diff
@@ -1,6 +1,17 @@
+    pass
+
+
+class BigQuerySourceNotFoundException(DataSourceNotFoundException):
```
I'm not sure if it's a good idea to have implementation specific errors. We are moving to having all implementations be self-contained. Any downside to using `DataSourceNotFoundException` instead of subclassing?

codyjlin replied:
Ah thanks for letting me know - I was following the example of the `FeastObjectNotFoundException` subclassing pattern [here]( https://github.com/feast-dev/feast/blob/master/sdk/python/feast/errors.py#L6-L33), but since the `DataSourceNotFoundException`s are so similar (and we're not checking `FileSource` yet), I'll just change it to use `DataSourceNotFoundException` instead.

woop commented:
/lgtm

tsotnet commented:
> FYI @tsotnet is working on updating the tests on our end.
> 
> @tsotnet , we need to update any feature repo in our tests that references the `rh_prod` gcp project to something valid, preferrably from https://cloud.google.com/bigquery/public-data

In the medium term, we'll have BigQuery tables created by our data, so we can run meaningful tests on the data across offline stores. But that won't land for a while, so we could replace the tables with public ones. It's quite hacky since we're using the FeatureViews from this file to test the online store (with different data), but I'm okay with it since it's a temporary workaround.

tsotnet commented on sdk/python/feast/data_source.py:
```diff
@@ -670,6 +671,7 @@ def __init__(
         query: Optional[str] = None,
     ):
         self._bigquery_options = BigQueryOptions(table_ref=table_ref, query=query)
+        self._assert_table_exists()
```
External validations like these should not happen in the constructor. Can you move it to happen when the user runs `feast apply`?

codyjlin commented:
Note that with 2b81e00, this PR now depends on https://github.com/feast-dev/feast/issues/1645 (PR https://github.com/feast-dev/feast/pull/1646 by @mavysavydav) so proper checking and network requests all occur at apply time.

woop commented on sdk/python/tests/test_cli_gcp.py:
```diff
@@ -53,3 +53,36 @@ def test_basic() -> None:
+
+
+@pytest.mark.integration
+def test_fail() -> None:
```
Nitpick, but `test_fail` isn't descriptive. If this test fails in CI then we won't have a clue as to what went wrong. Can we use a more precise description so that we know what the scope of this test is?

codyjlin replied:
Yes definitely will do.

woop replied:
I think the actual function needs to be renamed btw.

woop commented on sdk/python/tests/example_feature_repo_3.py:
```diff
@@ -0,0 +1,20 @@
+from datetime import timedelta
```
Can we rename `example_feature_repo_3.py` to something more descriptive like `feature_repo_with_missing_bq_source.py`?

codyjlin replied:
Yes definitely, thanks for the feedback.

tsotnet commented on sdk/python/feast/repo_operations.py:
```diff
@@ -108,6 +109,17 @@ def parse_repo(repo_root: Path) -> ParsedRepo:
     return res
 
 
+def assert_table_exists(table_ref):
```
Let's move this to the data source and make it part of the public API:

1. Add `validate()` method to `DataSource` class that raises `NotImplementedError`.
2. Override the method in `BigQuerySource` and move the contents of this function to that method
3. Override the method in `FileSource` without doing any validations there for now (we can validate that the file exists locally, in s3, or in gs in a separate PR)

tsotnet replied:
Also, let's move the conditional check from below (`not data_source.query`) here.

tsotnet commented on sdk/python/feast/repo_operations.py:
```diff
@@ -156,6 +168,9 @@ def apply_total(repo_config: RepoConfig, repo_path: Path):
             repo_config.offline_store, data_source
         )
 
+        if isinstance(data_source, BigQuerySource) and not data_source.query:
+            assert_table_exists(data_source.table_ref)
```
Let's get rid of the if statement and call `data_source.validate()` here.

woop replied:
btw, isn't it more Pythonic to do `is_valid()`? https://docs.python.org/3/library/stdtypes.html?highlight=identifier#str.isidentifier

codyjlin replied:
Perhaps it is standard for `is_<conditional>()` to return True or False, but `validate()` is ok since it makes assertions and would result in an exception on false? I'm open to either.

codyjlin commented:
Rebased and was able to fix the tests (at least all of the local ones). Thanks for all the comments and help @tsotnet @achals and @woop!

codyjlin commented:
From https://github.com/codyjlin/feast/actions/runs/951680702, it looks like a few tests still failed because they now need the `bigquery.Client()` if they're trying to register a BigQueryDataSource. The errors are all missing `GOOGLE_APPLICATION_CREDENTIALS` - can I just label these tests with `@pytest.mark.integration`?

achals commented:
> From https://github.com/codyjlin/feast/actions/runs/951680702, it looks like a few tests still failed because they now need the `bigquery.Client()` if they're trying to register a BigQueryDataSource. The errors are all missing `GOOGLE_APPLICATION_CREDENTIALS` - can I just label these tests with `@pytest.mark.integration`?

Yes, I think that's fine!

achals reviewed (approved):
/lgtm
